### PR TITLE
101 idlcache

### LINKED ISSUE #101: idl/entry cache layer
To prevent as much work deserialising entries/idl's, we should be able to put a cache layer between the be::mod and be::idlayer. Note this relies on #8 to be completed. The caching system itself is still a bit up for debate due to needing MVCC properties, so we probably need the tree from concread to be complete first.

### PR DESCRIPTION
**Pull request by Firstyear**
Fixes #101, concurrent caching of IDL and Entries. This yields a 10% improvement for test case execution, and 35% for tests run under --release mode. A lot of code around the code base was needed to be touched due to the extra need for mut in some operations and some lifetimes, but the majority of the work was in idl_arc_sqlite.rs, which has the cache layer. There are many performance gains yet to see, but most of those will come through improvement of the concread ARC and it's related BTree implementation. 

- [ x ] cargo fmt has been run
- [ x ] cargo test has been run and passes
- [ - ] design document included (if relevant)

### CODE REVIEW

File: kanidmd/src/lib/actors/v1_read.rs [hunk 1]
```diff
@@ -221,9 +221,8 @@ impl Handler<SearchMessage> for QueryServerReadV1 {
             audit_log!(audit, "Begin event {:?}", srch);
 
             match qs_read.search_ext(&mut audit, &srch) {
```
[the remark this answers is not in the document]
Comment (Firstyear) [reply kanidmd/src/lib/actors/v1_read.rs:223]: Yeah, this looks like an Older part of the code, before ? was something I knew how to use. There is probably a lot in the code base that could have this cleanup. I'd rather do it as part of #195, so I took a note there.

File: kanidmd/src/lib/be/idl_arc_sqlite.rs [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+        self.db.write_identries_raw(au, entries)
+    }
+
+    pub fn delete_identry<I>(&mut self, au: &mut AuditScope, idl: I) -> Result<(), OperationError>
+    where
+        I: Iterator<Item = u64>,
+    {
+        // Danger! We know that the entry cache is valid to manipulate here
+        // but rust doesn't know that so it prevents the mut/immut borrow.
+        let e_cache = unsafe { &mut *(&mut self.entry_cache as *mut ArcWriteTxn<_, _>) };
```
[the remark this answers is not in the document]
Comment (Firstyear) [reply kanidmd/src/lib/be/idl_arc_sqlite.rs:345]: Yes, Arc does all the copy-on-write. In this case, Rust won't let you pass the self.cache to the closure as that mutable borrow prevents the self.db.delete_identry. We know that self.db won't mut anything, and self.entry_cache MUST persistnt for the whole operation, and it internally does it's mutation without rewriting any root pointers etc. So as a result, this is safe, but we just need to tell rust that that condition is true.

File: kanidmd/src/lib/be/idl_arc_sqlite.rs [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+    pub fn setup(&self, audit: &mut AuditScope) -> Result<(), OperationError> {
+        self.db.setup(audit)
+    }
+}
+
+impl IdlArcSqlite {
+    pub fn new(audit: &mut AuditScope, path: &str, pool_size: u32) -> Result<Self, OperationError> {
+        let db = IdlSqlite::new(audit, path, pool_size)?;
+        let entry_cache = Arc::new(
+            DEFAULT_CACHE_TARGET,
```
[the remark this answers is not in the document]
Comment (Firstyear) [reply kanidmd/src/lib/be/idl_arc_sqlite.rs:440]: Later yes, for now no. I haven't thought of how to do the configuration for this yet .... today it's really only me running it so we can easily tweak the config/tuning later as the bulk of the hard work is done, and the wiring between cli config -> be is already in place, so it won't be hard to add as required. And that gives time to think about how I want to handle configuration in the future too ....

File: kanidmd/src/lib/actors/v1_write.rs [hunk 4]
```diff
@@ -497,14 +499,18 @@ impl Handler<ReviveRecycledMessage> for QueryServerWriteV1 {
...
-                        audit_log!(audit, "Failed to begin revive: {:?}", e);
-                        return Err(e);
-                    }
-                };
+            let rev = match ReviveRecycledEvent::from_parts(
+                &mut audit,
+                msg.uat,
+                msg.filter,
+                &mut qs_write,
+            ) {
```
[the remark this answers is not in the document]
Comment (Firstyear) [reply kanidmd/src/lib/actors/v1_write.rs:507]: As above, I think this is #195

File: kanidmd/src/lib/actors/v1_read.rs [hunk 5]
```diff
@@ -669,15 +669,18 @@ impl Handler<InternalSshKeyTagReadMessage> for QueryServerReadV1 {
...
-                        audit_log!(audit, "Failed to begin search: {:?}", e);
-                        return Err(e);
-                    }
-                };
+            let srch = match SearchEvent::from_target_uuid_request(
+                &mut audit,
+                uat,
+                target_uuid,
+                &mut qs_read,
+            ) {
```
[the remark this answers is not in the document]
Comment (Firstyear) [reply kanidmd/src/lib/actors/v1_read.rs:677]: Also #195

File: kanidmd/src/lib/actors/v1_read.rs [hunk 6]
```diff
@@ -373,21 +372,22 @@ impl Handler<InternalSearchRecycledMessage> for QueryServerReadV1 {
...
             // Make an event from the request
-            let srch = match SearchEvent::from_internal_recycle_message(&mut audit, msg, &qs_read) {
-                Ok(s) => s,
-                Err(e) => {
-                    audit_log!(audit, "Failed to begin recycled search: {:?}", e);
-                    return Err(e);
-                }
-            };
+            let srch =
+                match SearchEvent::from_internal_recycle_message(&mut audit, msg, &mut qs_read) {
```
[the remark this answers is not in the document]
Comment (Firstyear) [reply kanidmd/src/lib/actors/v1_read.rs:379]: #195

File: kanidmd/src/lib/be/idl_arc_sqlite.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (pando85) [reply kanidmd/src/lib/be/idl_arc_sqlite.rs:440]: Ok, great!

File: kanidmd/src/lib/actors/v1_read.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (pando85) [reply kanidmd/src/lib/actors/v1_read.rs:223]: cool

**Pull request merged**